core/src/main/scala/cats/Selective.scala

```diff
@@ -0,0 +1,86 @@
+package cats
+
+import simulacrum.{noop, typeclass}
+import scala.annotation.implicitNotFound
+
+@implicitNotFound("Could not find an instance of Selective for ${F}")
+@typeclass trait Selective[F[_]] extends Applicative[F] {
+  def select[A, B](fab: F[Either[A, B]])(ff: F[A => B]): F[B]
+
+  def branch[A, B, C](fab: F[Either[A, B]])(fl: F[A => C])(fr: F[B => C]): F[C] = {
+    val innerLhs: F[Either[A, Either[B, C]]] = map(fab)(_.map(Left(_)))
+    val innerRhs: F[A => Either[B, C]] = map(fl)(_.andThen(Right(_)))
+    val lhs = select(innerLhs)(innerRhs)
+    select(lhs)(fr)
+  }
+
+  def apS[A, B](ff: F[A => B])(fa: F[A]): F[B] = {
```

Comment:
Why do we want this?

Reply:
Parroting a law.  I think we agree the `ff` should be lazy on `select`, but I don't see an opportunity for laziness in either parameter here.

It avoids `flatMap` on `FlatMap`s, which may be faster for `Parser` and more amenable to static analysis.  But if that's the rationale, where does it end? `productS`? `map2S`? `productRS`?

Reply:
Wow. I didn't notice similar was being discussed in #3712 today!

Reply:
Got rid of this by caling it `ap` and making it a law on `RigidSelective`.